p2panda-net-next/src/actors/gossip/mod.rs

```diff
@@ -0,0 +1,987 @@
+// SPDX-License-Identifier: MIT OR Apache-2.0
+
+//! An `iroh`-specific gossip actor for message broadcast.
+
+mod joiner;
+mod listener;
+mod receiver;
+mod sender;
+mod session;
+
+use std::collections::{HashMap, HashSet};
+
+use iroh::Endpoint as IrohEndpoint;
+use iroh::NodeId;
+use iroh_gossip::net::Gossip as IrohGossip;
+use iroh_gossip::proto::{Config as IrohGossipConfig, DeliveryScope as IrohDeliveryScope};
+use p2panda_core::PublicKey;
+use ractor::{
+    Actor, ActorId, ActorProcessingErr, ActorRef, Message, RpcReplyPort, SupervisionEvent,
+};
+use tokio::sync::mpsc::{self, Receiver, Sender};
+use tokio::sync::oneshot::{self, Sender as OneshotSender};
+use tracing::{debug, warn};
+
+use crate::actors::gossip::session::{GossipSession, ToGossipSession};
+use crate::network::{FromNetwork, ToNetwork};
+use crate::{TopicId, from_public_key};
+
+pub enum ToGossip {
+    /// Return a handle to the iroh gossip actor.
+    ///
+    /// This is required when registering the gossip ALPN with the router.
+    Handle(RpcReplyPort<IrohGossip>),
+
+    /// Subscribe to the given topic, using the given peers as gossip bootstrap nodes.
+    Subscribe(
+        TopicId,
+        Vec<PublicKey>,
+        RpcReplyPort<(Sender<ToNetwork>, Receiver<FromNetwork>)>,
+    ),
+
+    /// Join a set of peers on the given gossip topic.
+    ///
+    /// This event requires a prior subscription to the topic via the `ToGossip::Subscribe`.
+    JoinPeers(TopicId, Vec<PublicKey>),
+
+    /// Joined a topic by connecting to the given peers.
+    Joined {
+        topic_id: TopicId,
+        peers: Vec<PublicKey>,
+        session_id: ActorId,
+    },
+
+    /// Gained a new, direct neighbor in the gossip overlay.
+    NeighborUp {
+        peer: PublicKey,
+        session_id: ActorId,
+    },
+
+    /// Lost a direct neighbor in the gossip overlay.
+    NeighborDown {
+        peer: PublicKey,
+        session_id: ActorId,
+    },
+
+    /// Received a message from the gossip overlay.
+    ReceivedMessage {
+        bytes: Vec<u8>,
+        delivered_from: PublicKey,
+        delivery_scope: IrohDeliveryScope,
+        topic_id: TopicId,
+        session_id: ActorId,
+    },
+}
+
+impl Message for ToGossip {}
+
+pub struct GossipState {
+    gossip: IrohGossip,
+    sessions_by_actor_id: HashMap<ActorId, TopicId>,
+    sessions_by_topic_id: HashMap<TopicId, Vec<ActorRef<ToGossipSession>>>,
+    neighours_by_topic_id: HashMap<TopicId, HashSet<PublicKey>>,
+    from_gossip_senders: HashMap<TopicId, Vec<Sender<FromNetwork>>>,
+    gossip_joined_senders: HashMap<ActorId, OneshotSender<u8>>,
+    topic_delivery_scopes: HashMap<TopicId, Vec<IrohDeliveryScope>>,
+}
```

Comment:
Maybe good to have some assertions in the tests just to make sure these maps really contain what we expect after state transitions